src/Datadog.Trace.ClrProfiler.Managed/CallTarget/Handlers/IntegrationMapper.cs

```diff
@@ -0,0 +1,714 @@
+                instanceProxyType = result.ProxyType;
+                callGenericTypes.Add(instanceProxyType);
+            }
+            else
+            {
+                callGenericTypes.Add(targetType);
+            }
+
+            DynamicMethod callMethod = new DynamicMethod(
+                     $"{onMethodEndMethodInfo.DeclaringType.Name}.{onMethodEndMethodInfo.Name}",
+                     typeof(CallTargetReturn),
+                     new Type[] { targetType, typeof(Exception), typeof(CallTargetState) },
+                     onMethodEndMethodInfo.Module,
+                     true);
+
+            ILGenerator ilWriter = callMethod.GetILGenerator();
+
+            // Load the instance if is needed
+            if (mustLoadInstance)
+            {
+                ilWriter.Emit(OpCodes.Ldarg_0);
+
+                if (instanceGenericConstraint != null)
+                {
+                    ConstructorInfo ctor = instanceProxyType.GetConstructors()[0];
+                    if (targetType.IsValueType && !ctor.GetParameters()[0].ParameterType.IsValueType)
+                    {
+                        ilWriter.Emit(OpCodes.Box, targetType);
+                    }
+
+                    ilWriter.Emit(OpCodes.Newobj, ctor);
+                }
+            }
+
+            // Load the exception
+            ilWriter.Emit(OpCodes.Ldarg_1);
+
+            // Load the state
+            ilWriter.Emit(OpCodes.Ldarg_2);
+
+            // Call Method
+            onMethodEndMethodInfo = onMethodEndMethodInfo.MakeGenericMethod(callGenericTypes.ToArray());
+            ilWriter.EmitCall(OpCodes.Call, onMethodEndMethodInfo, null);
+
+            ilWriter.Emit(OpCodes.Ret);
+
+            Log.Debug($"Created EndMethod Dynamic Method for '{integrationType.FullName}' integration. [Target={targetType.FullName}]");
+            return callMethod;
+        }
+
+        internal static DynamicMethod CreateEndMethodDelegate(Type integrationType, Type targetType, Type returnType)
+        {
+            Log.Debug($"Creating EndMethod Dynamic Method for '{integrationType.FullName}' integration. [Target={targetType.FullName}, ReturnType={returnType.FullName}]");
+            MethodInfo onMethodEndMethodInfo = integrationType.GetMethod(EndMethodName, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static);
+            if (onMethodEndMethodInfo is null)
+            {
+                return null;
+            }
+
+            if (onMethodEndMethodInfo.ReturnType.GetGenericTypeDefinition() != typeof(CallTargetReturn<>))
+            {
+                throw new ArgumentException($"The return type of the method: {EndMethodName} in type: {integrationType.FullName} is not {nameof(CallTargetReturn)}");
+            }
+
+            Type[] genericArgumentsTypes = onMethodEndMethodInfo.GetGenericArguments();
+            if (genericArgumentsTypes.Length < 1 || genericArgumentsTypes.Length > 2)
```

Comment:
This check means that `genericArgumentsTypes.Length` may be either 1 or 2. I'm trying to understand in what scenarios it may be 1...the only one I can come up with is if the return type is a common type like `int` or `string` or `object` so that we only require `TTarget`. Is that correct? Are there any other scenarios I'm missing?

Reply:
No, you are correct! If we have access to the return type at compile time, we don't need the constraint or ducktyping, we should be able just to put the type there.

Reply:
Thanks for the clarification! I'll leave this as unresolved in case it helps other reviewers to have this additional context